### DISCUSSION TITLE
how does meson generate a build file when fetching a subproject with a .wrap file

### DISCUSSION THREAD

**Discussion by BruceLi686** [mesonbuild/meson#12343] [Q&A] [answered]
In a meson project, i found it seems to automatically generate a meson.build file for a downloaded subproject according to the subproject's .wrap file. Is it true? How does meson achieve that? I'm a newbie to meson, could anyone guide me?

<img width="200" alt="image" src="https://github.com/mesonbuild/meson/assets/73571819/c2518fce-9b80-4669-8d66-fb3617c2aeca">

**Comment by QuLogic** [comment]
There is no automatic generation; that project has a `meson.build`: https://github.com/nlohmann/json/blob/develop/meson.build

**Reply by BruceLi686** [reply]
I take a bad example. Below is a new example to illustrate my question.
<img width="358" alt="image" src="https://github.com/mesonbuild/meson/assets/73571819/152cae38-f038-4e8a-af5b-17cb0331d95e">
In original socket.io-client-cpp.git, no meson.build file. But you can note there's a 'patch_directory' entry in .wrap file. Then meson will copy the meson.build file from **sioclient** in **packagefiles**. This is the overlay mechanism of meson, details see https://mesonbuild.com/Wrap-dependency-system-manual.html#accepted-configuration-properties-for-wraps .

Thanks @QuLogic  for your reply.

**Comment by eli-schwartz** [comment, marked as answer]
> it seems to automatically generate a meson.build file for a downloaded subproject according to the subproject's .wrap file. Is it true? How does meson achieve that?

No. A wrap file indicates that the *entire* source code for the subproject is downloaded from elsewhere. This includes both the `.hpp` / `.cpp` / '.h' / '.c' files (and others) as well as the meson.build itself. In cases where the upstream project doesn't have a meson.build file in their source code tarball, a .wrap file may specify two archives to download, one of which contains the upstream source code and the other containing the meson.build file(s).

**Reply by BruceLi686** [reply]
Yes. The .wrap file defines a patch_directory entry, then any files/archive placed in subprojects/packagefiles/%patch_directory% will be overlayed.

Thank you 😊 @eli-schwartz .
